glean-core/ios/sdk_generator.sh

```diff
@@ -0,0 +1,84 @@
+#!/bin/bash
+
+# This Source Code Form is subject to the terms of the Mozilla Public
+# License, v. 2.0. If a copy of the MPL was not distributed with this
+# file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+# Glean SDK metrics build script.
+#
+# More about Glean at https://mozilla.github.io/glean
+#
+# This script generates metrics and pings as defined in user-provided files
+# and generates Swift code to be included in the final build.
+# It uses the `glean_parser`.
+# See https://mozilla.github.io/glean_parser/ for details.
+#
+# To use it in a Swift project, follow these steps:
+# 1. Import the `sdk_generator.sh` script into your project.
+# 2. Add your `metrics.yaml` and (optionally) `pings.yaml` to your project.
+# 3. Add a new "Run Script" build step and set the command to `bash $PWD/sdk_generator.sh`
+# 4. Add your definition files (`metrics.yaml`, `pings.yaml`) as Input Files for the "Run Script" step.
+# 5. Run the build.
+# 6. Add the files in the `Generated` folder to your project.
+# 7. Add the same files from the `Generated` folder as Output Files of the newly created "Run SCript" step.
+# 8. Start using the generated metrics.
+
+set -e
+
+GLEAN_PARSER_VERSION=1.8.2
+
+# This script runs a given Python module as a "main" module, like
+# `python -m module`. However, it first checks that the installed
+# package is at the desired version, and if not, upgrades it using `pip`.
+RUN_PYTHON_SCRIPT=$(cat <<'EOF'
+import importlib
+import subprocess
+import sys
+
+module_name = sys.argv[1]
+expected_version = sys.argv[2]
+
+try:
+    module = importlib.import_module(module_name)
+except ImportError:
+    found_version = None
+else:
+    found_version = getattr(module, '__version__')
+
+if found_version != expected_version:
+    subprocess.check_call([
+        sys.executable,
+        '-m',
+        'pip',
+        'install',
+        '--upgrade',
+        f'{module_name}=={expected_version}'
+    ])
+
+subprocess.check_call([
+    sys.executable,
+    '-m',
+    module_name
+] + sys.argv[3:])
+EOF
+)
```

Comment:
This whole thing is really a hack to get around limitations in Gradle and being unable to call `pip` in the normal way.  Since here we're just writing a bash script, we can just do (from bash, not this Python stub):

```
~/.local/glean/venv/bin/pip install --upgrade glean_parser==$GLEAN_PARSER_VERSION
~/.local/glean/venv/bin/python3 -m glean_parser translate ...
```

(And should probably put the path to the central venv in a variable, too...)

Reply:
Uhm, that makes _a lot more_ sense.